File: problems/java/ParkingLotSystem/src/builders/ParkingLotBuilder.java
package builders;

import enums.VehicleSize;
import model.ParkingFloor;
import service.ParkingLotSystem;

import java.util.ArrayList;
import java.util.List;

/**
 * Builder for creating a parking lot structure (floors and spots) without
 * exposing ParkingFloor / ParkingSpot construction to the client.
 * Client uses this API instead of instantiating model classes directly.
 */
public class ParkingLotBuilder {

    private final List<ParkingFloor> floors = new ArrayList<>();
    private ParkingFloor currentFloor;

    /**
     * Adds a new floor with the given number. Subsequent addSpot() calls
     * will add spots to this floor until the next addFloor().
     */
    public ParkingLotBuilder addFloor(int floorNumber) {
        currentFloor = new ParkingFloor(floorNumber);
        floors.add(currentFloor);
        return this;
    }

    /**
     * Adds a spot to the current floor. Must call addFloor() first.
     */
    public ParkingLotBuilder addSpot(VehicleSize size, String spotId) {
        if (currentFloor == null) {
            throw new IllegalStateException("Call addFloor() before addSpot()");
        }
        currentFloor.addSpot(size, spotId);
        return this;
    }

    /**
     * Returns the built list of floors. Client can pass these to ParkingLotSystem.addFloor().
     */
    public List<ParkingFloor> build() {
        return new ArrayList<>(floors);
    }

    /**
     * Builds and adds all floors to the system. Client does not need to reference ParkingFloor or List.
     */
    public void buildAndAddTo(ParkingLotSystem system) {
        build().forEach(system::addFloor);
    }
}

File: problems/python/ParkingLotSystem/src/builders/parking_lot_builder.py
"""Builder for parking lot floors and spots."""
from __future__ import annotations
from typing import Optional
from enums.vehicle_size import VehicleSize
from model.parking_floor import ParkingFloor
from service.parking_lot_system import ParkingLotSystem


class ParkingLotBuilder:
    def __init__(self):
        self._floors: list[ParkingFloor] = []
        self._current_floor: Optional[ParkingFloor] = None

    def add_floor(self, floor_number: int) -> "ParkingLotBuilder":
        self._current_floor = ParkingFloor(floor_number)
        self._floors.append(self._current_floor)
        return self

    def add_spot(self, size: VehicleSize, spot_id: str) -> "ParkingLotBuilder":
        if self._current_floor is None:
            raise RuntimeError("Call add_floor() before add_spot()")
        self._current_floor.add_spot(size, spot_id)
        return self

    def build(self) -> list[ParkingFloor]:
        return list(self._floors)

    def build_and_add_to(self, system: ParkingLotSystem) -> None:
        for floor in self.build():
            system.add_floor(floor)
